File: frontend/src/services/wateringService.ts
/**
 * API service for watering operations
 */
import {
  WateringSchedule,
  WateringScheduleCreate,
  WateringScheduleUpdate,
  WateringScheduleWithNextDate,
  WateringLog,
  WateringLogCreate,
} from '@/types/watering'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000'

export const wateringService = {
  // Schedule operations
  async getSchedule(scheduleId: string): Promise<WateringSchedule> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/schedules/${scheduleId}`
    )
    if (!response.ok) {
      throw new Error('Failed to fetch watering schedule')
    }
    return response.json()
  },

  async getScheduleWithNextDate(
    scheduleId: string
  ): Promise<WateringScheduleWithNextDate> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/schedules/${scheduleId}/next-date`
    )
    if (!response.ok) {
      throw new Error('Failed to fetch watering schedule with next date')
    }
    return response.json()
  },

  async getPlantSchedules(
    plantId: string,
    activeOnly: boolean = false
  ): Promise<WateringSchedule[]> {
    const params = new URLSearchParams()
    if (activeOnly) {
      params.append('active_only', 'true')
    }
    const url = `${API_BASE_URL}/api/v1/watering/plants/${plantId}/schedules${
      params.toString() ? `?${params.toString()}` : ''
    }`
    const response = await fetch(url)
    if (!response.ok) {
      throw new Error('Failed to fetch plant watering schedules')
    }
    return response.json()
  },

  async createSchedule(
    plantId: string,
    scheduleData: Omit<WateringScheduleCreate, 'plant_id'>
  ): Promise<WateringSchedule> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/plants/${plantId}/schedules`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...scheduleData,
          plant_id: plantId,
        }),
      }
    )
    if (!response.ok) {
      throw new Error('Failed to create watering schedule')
    }
    return response.json()
  },

  async updateSchedule(
    scheduleId: string,
    scheduleData: WateringScheduleUpdate
  ): Promise<WateringSchedule> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/schedules/${scheduleId}`,
      {
        method: 'PUT',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify(scheduleData),
      }
    )
    if (!response.ok) {
      throw new Error('Failed to update watering schedule')
    }
    return response.json()
  },

  async deleteSchedule(scheduleId: string): Promise<void> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/schedules/${scheduleId}`,
      {
        method: 'DELETE',
      }
    )
    if (!response.ok) {
      throw new Error('Failed to delete watering schedule')
    }
  },

  // Log operations
  async getPlantLogs(plantId: string, limit: number = 50): Promise<WateringLog[]> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/plants/${plantId}/logs?limit=${limit}`
    )
    if (!response.ok) {
      throw new Error('Failed to fetch watering logs')
    }
    return response.json()
  },

  async createLog(
    plantId: string,
    logData: Omit<WateringLogCreate, 'plant_id'>
  ): Promise<WateringLog> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/plants/${plantId}/logs`,
      {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
        },
        body: JSON.stringify({
          ...logData,
          plant_id: plantId,
        }),
      }
    )
    if (!response.ok) {
      throw new Error('Failed to create watering log')
    }
    return response.json()
  },

  async deleteLog(logId: string): Promise<void> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/logs/${logId}`,
      {
        method: 'DELETE',
      }
    )
    if (!response.ok) {
      throw new Error('Failed to delete watering log')
    }
  },

  // Due watering
  async getDueForWatering(
    daysAhead: number = 0
  ): Promise<WateringScheduleWithNextDate[]> {
    const response = await fetch(
      `${API_BASE_URL}/api/v1/watering/due?days_ahead=${daysAhead}`
    )
    if (!response.ok) {
      throw new Error('Failed to fetch due watering schedules')
    }
    return response.json()
  },
}

File: backend/app/services/watering_service.py
"""Service for watering operations."""
from uuid import UUID
from datetime import date

from fastapi import HTTPException
from sqlalchemy.ext.asyncio import AsyncSession

from app.repositories.watering_repository import WateringRepository
from app.repositories.plant_repository import PlantRepository
from app.schemas.watering import (
    WateringScheduleCreate,
    WateringScheduleUpdate,
    WateringScheduleResponse,
    WateringScheduleWithNextDate,
    WateringLogCreate,
    WateringLogResponse,
)


class WateringService:
    """Service for watering business logic."""

    def __init__(self, db: AsyncSession):
        """Initialize the service."""
        self.db = db
        self.watering_repo = WateringRepository(db)
        self.plant_repo = PlantRepository(db)

    # Watering Schedule Methods
    async def get_schedule_by_id(
        self, schedule_id: UUID
    ) -> WateringScheduleResponse:
        """Get a watering schedule by ID."""
        schedule = await self.watering_repo.get_schedule_by_id(schedule_id)
        if not schedule:
            raise HTTPException(status_code=404, detail="Watering schedule not found")
        return WateringScheduleResponse.model_validate(schedule)

    async def get_schedules_by_plant_id(
        self, plant_id: UUID, active_only: bool = False
    ) -> list[WateringScheduleResponse]:
        """Get all watering schedules for a plant."""
        # Verify plant exists
        plant = await self.plant_repo.get_by_id(plant_id)
        if not plant:
            raise HTTPException(status_code=404, detail="Plant not found")

        schedules = await self.watering_repo.get_schedules_by_plant_id(
            plant_id, active_only
        )
        return [WateringScheduleResponse.model_validate(s) for s in schedules]

    async def create_schedule(
        self, schedule_data: WateringScheduleCreate
    ) -> WateringScheduleResponse:
        """Create a new watering schedule."""
        # Verify plant exists
        plant = await self.plant_repo.get_by_id(schedule_data.plant_id)
        if not plant:
            raise HTTPException(status_code=404, detail="Plant not found")

        # Validate date range
        if schedule_data.end_date and schedule_data.end_date < schedule_data.start_date:
            raise HTTPException(
                status_code=400, detail="End date must be after start date"
            )

        schedule = await self.watering_repo.create_schedule(schedule_data)
        return WateringScheduleResponse.model_validate(schedule)

    async def update_schedule(
        self, schedule_id: UUID, schedule_data: WateringScheduleUpdate
    ) -> WateringScheduleResponse:
        """Update a watering schedule."""
        # Get existing schedule to validate
        existing = await self.watering_repo.get_schedule_by_id(schedule_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Watering schedule not found")

        # Validate date range if both dates are provided
        start_date = schedule_data.start_date or existing.start_date
        end_date = schedule_data.end_date if schedule_data.end_date is not None else existing.end_date

        if end_date and end_date < start_date:
            raise HTTPException(
                status_code=400, detail="End date must be after start date"
            )

        schedule = await self.watering_repo.update_schedule(schedule_id, schedule_data)
        if not schedule:
            raise HTTPException(status_code=404, detail="Watering schedule not found")

        return WateringScheduleResponse.model_validate(schedule)

    async def delete_schedule(self, schedule_id: UUID) -> None:
        """Delete a watering schedule."""
        success = await self.watering_repo.delete_schedule(schedule_id)
        if not success:
            raise HTTPException(status_code=404, detail="Watering schedule not found")

    # Watering Log Methods
    async def get_logs_by_plant_id(
        self, plant_id: UUID, limit: int = 50
    ) -> list[WateringLogResponse]:
        """Get watering logs for a plant."""
        # Verify plant exists
        plant = await self.plant_repo.get_by_id(plant_id)
        if not plant:
            raise HTTPException(status_code=404, detail="Plant not found")

        logs = await self.watering_repo.get_logs_by_plant_id(plant_id, limit)
        return [WateringLogResponse.model_validate(log) for log in logs]

    async def create_log(self, log_data: WateringLogCreate) -> WateringLogResponse:
        """Create a new watering log entry."""
        # Verify plant exists
        plant = await self.plant_repo.get_by_id(log_data.plant_id)
        if not plant:
            raise HTTPException(status_code=404, detail="Plant not found")

        # Verify schedule exists if provided
        if log_data.watering_schedule_id:
            schedule = await self.watering_repo.get_schedule_by_id(
                log_data.watering_schedule_id
            )
            if not schedule:
                raise HTTPException(
                    status_code=404, detail="Watering schedule not found"
                )

        log = await self.watering_repo.create_log(log_data)
        return WateringLogResponse.model_validate(log)

    async def delete_log(self, log_id: UUID) -> None:
        """Delete a watering log."""
        success = await self.watering_repo.delete_log(log_id)
        if not success:
            raise HTTPException(status_code=404, detail="Watering log not found")

    # Advanced Methods
    async def get_schedule_with_next_date(
        self, schedule_id: UUID
    ) -> WateringScheduleWithNextDate:
        """Get a watering schedule with calculated next watering date."""
        schedule = await self.watering_repo.get_schedule_by_id(schedule_id)
        if not schedule:
            raise HTTPException(status_code=404, detail="Watering schedule not found")

        next_date = await self.watering_repo.calculate_next_watering_date(schedule)

        response = WateringScheduleWithNextDate.model_validate(schedule)
        response.next_watering_date = next_date
        return response

    async def get_plants_due_for_watering(
        self, days_ahead: int = 0
    ) -> list[WateringScheduleWithNextDate]:
        """Get plants that are due for watering."""
        due_plants = await self.watering_repo.get_plants_due_for_watering(days_ahead)

        results = []
        for schedule, next_date in due_plants:
            response = WateringScheduleWithNextDate.model_validate(schedule)
            response.next_watering_date = next_date
            results.append(response)

        return results
